libs/coin-modules/coin-mina/src/bridge/synchronisation.ts

```diff
@@ -0,0 +1,185 @@
+          invariant(op.metadata?.delegate_change_target, "mina: missing delegate change target");
+          toAccount = op.metadata?.delegate_change_target;
+          redelegateTransaction = true;
+          continue;
+        }
+        case "account_creation_fee_via_payment": {
+          accountCreationFee = opValue.times(-1);
+          continue;
+        }
+      }
+    }
+
+    invariant(fromAccount, "mina: missing fromAccount");
+    invariant(toAccount, "mina: missing toAccount");
+
+    const op: MinaOperation = {
+      id: "",
+      type: "NONE",
+      hash,
+      value,
+      fee,
+      blockHeight,
+      hasFailed: failed,
+      blockHash,
+      accountId,
+      senders: [fromAccount],
+      recipients: [toAccount],
+      date,
+      extra: {
+        memo,
+        accountCreationFee: accountCreationFee.toString(),
+      },
+    };
+
+    const ops: MinaOperation[] = [];
+    if (isSending) {
+      const type = "OUT";
+      ops.push({
+        ...op,
+        value: value.minus(accountCreationFee).plus(fee),
+        type,
+        id: encodeOperationId(accountId, hash, type),
+      });
+    } else if (redelegateTransaction) {
+      // delegate change
+      const type = "REDELEGATE";
+      ops.push({
+        ...op,
+        type,
+        id: encodeOperationId(accountId, hash, type),
+      });
+    } else {
+      const type = "IN";
+      ops.push({
+        ...op,
+        value: value.minus(accountCreationFee),
+        type,
+        id: encodeOperationId(accountId, hash, type),
+      });
+    }
+
+    return ops;
+  } catch (e) {
+    log("error", "mina: failed to convert txn to operation", {
+      error: e,
+      transaction: txn,
+    });
+    return [];
+  }
+};
+
+export const getAccountShape: GetAccountShape<MinaAccount> = async info => {
+  const { address, initialAccount, currency, derivationMode } = info;
+  const oldOperations = initialAccount?.operations || [];
+
+  const accountId = encodeAccountId({
+    type: "js",
+    version: "2",
+    currencyId: currency.id,
+    xpubOrAddress: address,
+    derivationMode,
+  });
+
+  const { blockHeight, balance, spendableBalance } = await getAccount(address);
+
+  const rosettaTxns = await getTransactions(address, initialAccount?.operations.length);
+  const newOperations = await Promise.all(
+    rosettaTxns.flatMap(t => mapRosettaTxnToOperation(accountId, address, t)),
+  );
+
+  const operations = mergeOps(oldOperations, newOperations.flat());
+
+  const shape: Partial<MinaAccount> = {
+    id: accountId,
+    balance,
+    spendableBalance,
+    operationsCount: operations.length,
+    blockHeight,
+  };
+
+  return { ...shape, operations };
+};
+
+const postSync = (initial: Account, synced: Account): Account => {
+  const pendingOperations = initial.pendingOperations || [];
+
+  if (pendingOperations.length === 0) {
+    return synced;
+  }
+
+  const { operations } = synced;
+
+  synced.pendingOperations = pendingOperations.filter(
```

Comment:
pendingOperations should be cleaned (discarded eg. array length becomes 0) automatically when it's returned by getAccountShape

When you broadcast, an optimistic operation is created and should have the right tx hash

Once that tx hash is added to operations array during the sync, the optimistic operation gets removed from pendingOperations array automatically

if you have an issue it means that either tx isn't present in the history fetching the second time, or tx hash is wrong in optimistic operation

Reply:
Ahh, I think it was a default method wrongly copied from some other place. Removed it!